client/json.go

```diff
@@ -0,0 +1,428 @@
+	return json.Marshal(v.val)
+}
+
+type jsonObject struct {
+	jsonBase[map[string]JSON]
+}
+
+var _ JSON = jsonObject{}
+
+func (obj jsonObject) Object() (map[string]JSON, bool) {
+	return obj.val, true
+}
+
+func (obj jsonObject) MarshalJSON() ([]byte, error) {
+	return json.Marshal(obj.val)
+}
+
+func (obj jsonObject) Unwrap() any {
+	result := make(map[string]any, len(obj.jsonBase.val))
+	for k, v := range obj.val {
+		result[k] = v.Unwrap()
+	}
+	return result
+}
+
+type jsonArray struct {
+	jsonBase[[]JSON]
+}
+
+var _ JSON = jsonArray{}
+
+func (arr jsonArray) Array() ([]JSON, bool) {
+	return arr.val, true
+}
+
+func (arr jsonArray) MarshalJSON() ([]byte, error) {
+	return json.Marshal(arr.val)
+}
+
+func (arr jsonArray) Unwrap() any {
+	result := make([]any, len(arr.jsonBase.val))
+	for i := range arr.val {
+		result[i] = arr.val[i].Unwrap()
+	}
+	return result
+}
+
+type jsonNumber struct {
+	jsonBase[float64]
+}
+
+var _ JSON = jsonNumber{}
+
+func (n jsonNumber) Number() (float64, bool) {
+	return n.val, true
+}
+
+func (n jsonNumber) MarshalJSON() ([]byte, error) {
+	return json.Marshal(n.val)
+}
+
+type jsonString struct {
+	jsonBase[string]
+}
+
+var _ JSON = jsonString{}
+
+func (s jsonString) String() (string, bool) {
+	return s.val, true
+}
+
+func (s jsonString) MarshalJSON() ([]byte, error) {
+	return json.Marshal(s.val)
+}
+
+type jsonBool struct {
+	jsonBase[bool]
+}
+
+var _ JSON = jsonBool{}
+
+func (b jsonBool) Bool() (bool, bool) {
+	return b.val, true
+}
+
+func (b jsonBool) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.val)
+}
+
+type jsonNull struct {
+	jsonVoid
+}
+
+var _ JSON = jsonNull{}
+
+func (n jsonNull) IsNull() bool {
+	return true
+}
+
+func (n jsonNull) Value() any {
+	return nil
+}
+
+func (n jsonNull) Unwrap() any {
+	return nil
+}
+
+func (n jsonNull) Marshal(w io.Writer) error {
+	return json.NewEncoder(w).Encode(nil)
+}
+
+func (n jsonNull) MarshalJSON() ([]byte, error) {
+	return json.Marshal(nil)
+}
+
+func newJSONObject(val map[string]JSON) JSON {
+	return jsonObject{jsonBase[map[string]JSON]{val: val}}
+}
+
+func newJSONArray(val []JSON) JSON {
+	return jsonArray{jsonBase[[]JSON]{val: val}}
+}
+
+func newJSONNumber(val float64) JSON {
+	return jsonNumber{jsonBase[float64]{val: val}}
+}
+
+func newJSONString(val string) JSON {
+	return jsonString{jsonBase[string]{val: val}}
+}
+
+func newJSONBool(val bool) JSON {
+	return jsonBool{jsonBase[bool]{val: val}}
+}
+
+func newJSONNull() JSON {
+	return jsonNull{}
+}
+
+// ParseJSONBytes parses the given JSON bytes into a JSON value.
+// Returns error if the input is not valid JSON.
+func ParseJSONBytes(data []byte) (JSON, error) {
+	var p fastjson.Parser
+	v, err := p.ParseBytes(data)
+	if err != nil {
+		return nil, err
+	}
+	return NewJSONFromFastJSON(v), nil
+}
+
+// ParseJSONString parses the given JSON string into a JSON value.
+// Returns error if the input is not valid JSON.
+func ParseJSONString(data string) (JSON, error) {
```

Comment:
I disagree. `ParseBytes` calls `Parse` internally so it would be weird to go from string to only to go back to string.

Reply:
Ah, well we can to the inverse then no?
```
func ParseJSONBytes(data string) (JSON, error) {
  return ParseJSONString([]byte(data)
}
```
The important part being that they are exactly the same/interchangable.

Reply:
Or if you'd argue that has the same problem and that it is the cast that you dislike - I disagree, as the cast is basically free and not much more than a compiler type-hint than anything serious at runtime, and in exchange we get a very clear guarantee and documentation that the two functions have the same behaviour.

Reply:
The thing is that it's more efficient to call the `ParseBytes` for bytes instead of casting bytes as string. `fastjson` uses an unsafe pointer to avoid copying memory which can be a lot more efficient for large JSON objects. If you absolutely want them to be the same we should do like they do.
```go
func PardeJSONBytes(data []byte) (JSON, error) {
    return ParseJSONString(*(*string)(unsafe.Pointer(&data)))
}
```

Reply:
Casting between `string` and `[]byte` copies in Go?  I'm happy as-is then :)

Reply:
> Casting between string and []byte copies in Go

yes, in both directions

Reply:
> That's what I meant by "fastjson.Parser.ParseBytes casts the bytes slice to a string internally"

Sorry - I got your comment notification after Fred's and his reply - I didn't mean to ignore you :)

Reply:
> yes, in both directions

Thanks!